### ISSUE TITLE
The Batch Normalization layer is not shown in tensorboard

### ISSUE THREAD

**Issue by lishihong-1** [ultralytics/ultralytics#12595]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I used tensorboard to view the model structure of yolov8s-ghost-p2, but the BN layer of the conv module is not shown in it though the BN layer is in the code. At the same time, I am not sure and want to know if the C3Ghost module uses the structure with stride=2, because I did not find a DWConv layer intensorboard.
The following is the official code:
<img width="682" alt="{515CBEEE-32CE-47a1-9E5B-EA7635DD0B9F}" src="https://github.com/ultralytics/ultralytics/assets/73936514/cd395bcc-b3fc-452e-b0d4-d8ec8e98cb28">
<img width="757" alt="{FC0C8CA1-D1CF-4677-A83C-9992BF562EE9}" src="https://github.com/ultralytics/ultralytics/assets/73936514/81fcd737-ba2d-410b-95ec-6dc7d9c26f33">

### Additional
<img width="653" alt="{DB630F32-FF39-4774-984D-D92EFB19952D}" src="https://github.com/ultralytics/ultralytics/assets/73936514/af49bc1c-4c4d-49b7-9435-30277a17336d">
<img width="671" alt="{BF1853F9-C488-4a5a-9ED3-ACE70A1C8BBC}" src="https://github.com/ultralytics/ultralytics/assets/73936514/1123f3a1-8d66-4e2d-807f-d607cedd216c">

_No response_

**Comment by glenn-jocher** [comment]
Hello! Thanks for bringing this issue to our attention. 

For the TensorBoard visualization, some elements like Batch Normalization (BN) may not be explicitly displayed due to how the model graph is processed and simplified in TensorBoard. These layers are typically integrated within the convolutional layers for performance optimization during inference.

Regarding the C3Ghost module with stride=2, it’s correct that TensorBoard may not show specific layers like DWConv if they are encapsulated within custom modules or implemented in ways that are optimized away in graphical outputs.

For a detailed internal structure visualization, I recommend checking the model's definition in the code or using more detailed visual tools that can parse these complex relationships better than TensorBoard.

Please let me know if there's anything else I can help with! 😊

**Comment by lishihong-1** [comment]
So when is the 'stride' parameter equal to 2 and when is it equal to 1 in the yolov8s-ghost-p2 model? I used the official .yaml file and didn't make any change.

**Comment by glenn-jocher** [comment]
Hello! In the YOLOv8s-ghost-p2 model, the 'stride' parameter is typically set to 2 in downsampling layers to reduce the spatial dimensions of the feature maps, effectively compressing the information and increasing the receptive field. Conversely, 'stride' is set to 1 in layers where spatial dimensions are preserved, which is usually done to maintain feature map size for detailed feature extraction.

This strategically alternates between preserving and compressing the information throughout the network. You can see these settings directly reflected in the .yaml configuration file under each layer's specifications without needing any modifications. Hope this clears things up! 😊

**Comment by lishihong-1** [comment]
However, when I review the structure of yolov8s-ghost-p2.pt in netron, I did not find the DW layer in the C3Ghost module's GhostBottleneck, and the.yaml file only write 'C3Ghost', so I cannot determine what kind of GhostBottleneck the model used. Also, I don't know why when I export the same model to onnx format, I didn't seem any model names like C3Ghost or SiLU, 
<img width="209" alt="{9C9CA950-7D51-4474-85A9-58CC88AD5F50}" src="https://github.com/ultralytics/ultralytics/assets/73936514/59318cff-699a-400c-b0d2-2d40d0d4c332">
<img width="343" alt="{300D33E1-64BD-43cc-9B8B-6A9818412E36}" src="https://github.com/ultralytics/ultralytics/assets/73936514/88157838-3c0f-4c8d-b7a1-aaba774dd96f">

**Comment by glenn-jocher** [comment]
Hi! The structure naming discrepancies you've observed between Netron and the `.yaml` config may be due to how these tools interpret and display the model layers. 

For example, specific layer names like `C3Ghost` or activation functions like `SiLU` might be represented differently in ONNX. This is because some naming conventions or modular structures might not carry over directly during the export process, and the names like `SiLU` might just appear as a generic activation layer.

Regarding not seeing the DW layer in the `GhostBottleneck` within the C3Ghost, it’s possible that it’s implemented in a manner that isn't distinctly separated as you would expect from the traditional layers configuration. It might help to look at the source code in the repository where `C3Ghost` and `GhostBottleneck` are defined to better understand the internal structure. The layers could be abstracted or integrated within other components in a way that Netron doesn't display straightforwardly.

If you can share the section of source code where the model and layers are defined, I might be able to provide more specific insights. 🙂

**Comment by lishihong-1** [comment]
class C3Ghost(C3):
    """C3 module with GhostBottleneck()."""

    def __init__(self, c1, c2, n=1, shortcut=True, g=1, e=0.5):
        """Initialize 'SPP' module with various pooling sizes for spatial pyramid pooling."""
        super().__init__(c1, c2, n, shortcut, g, e)
        c_ = int(c2 * e)  # hidden channels
        self.m = nn.Sequential(*(GhostBottleneck(c_, c_) for _ in range(n)))


class GhostBottleneck(nn.Module):
    """Ghost Bottleneck https://github.com/huawei-noah/ghostnet."""

    def __init__(self, c1, c2, k=3, s=1):
        """Initializes GhostBottleneck module with arguments ch_in, ch_out, kernel, stride."""
        super().__init__()
        c_ = c2 // 2
        self.conv = nn.Sequential(
            GhostConv(c1, c_, 1, 1),  # pw
            DWConv(c_, c_, k, s, act=False) if s == 2 else nn.Identity(),  # dw
            GhostConv(c_, c2, 1, 1, act=False),  # pw-linear
        )
        self.shortcut = (
            nn.Sequential(DWConv(c1, c1, k, s, act=False), Conv(c1, c2, 1, 1, act=False)) if s == 2 else nn.Identity()
        )

    def forward(self, x):
        """Applies skip connection and concatenation to input tensor."""
        return self.conv(x) + self.shortcut(x)

class GhostConv(nn.Module):
    """Ghost Convolution https://github.com/huawei-noah/ghostnet."""

    def __init__(self, c1, c2, k=1, s=1, g=1, act=True):
        """Initializes the GhostConv object with input channels, output channels, kernel size, stride, groups and
        activation.
        """
        super().__init__()
        c_ = c2 // 2  # hidden channels
        self.cv1 = Conv(c1, c_, k, s, None, g, act=act)
        self.cv2 = Conv(c_, c_, 5, 1, None, c_, act=act)

    def forward(self, x):
        """Forward propagation through a Ghost Bottleneck layer with skip connection."""
        y = self.cv1(x)
        return torch.cat((y, self.cv2(y)), 1)
This is the official code about GhostConv, GhostBottleNeck and C3Ghost module, I want to know about if the DWConv will be seen in netron or tensorboard according to this code?

**Comment by glenn-jocher** [comment]
Hey! 👋 Thanks for sharing the code snippets! Based on the code you provided, the DWConv (depth-wise convolution) layer in the `GhostBottleneck` is conditionally applied based on the stride (`s`). If the stride `s == 2`, then the DWConv layer is included in the sequence. Otherwise, an `nn.Identity()` layer is used, which means no operation is performed.

In tools like Netron or TensorBoard, whether you see the DWConv explicitly depends on the stride configuration used when visualizing the model. If layers are instantiated with a stride of 1, the DWConv layer essentially doesn't exist in that instance, hence won't be shown. For strides of 2, you should be able to see the DWConv in your Netron or TensorBoard visualizations. Keep in mind that the visualization might simply show the sequential block without expanding into details unless you delve deeper into the layer's operations. Hope this helps! 😊
